subsys/bluetooth/audio/bap_base.c

```diff
@@ -0,0 +1,562 @@
+
+			return NULL;
+		}
+
+		ltv_len = net_buf_simple_pull_u8(&net_buf);
+		if (net_buf.len < ltv_len) {
+			LOG_DBG("Invalid BASE length: %u", ad->data_len);
+
+			return NULL;
+		}
+		net_buf_simple_pull_mem(&net_buf, ltv_len);
+
+		/* Pull meta */
+		if (net_buf.len < sizeof(ltv_len)) {
+			LOG_DBG("Invalid BASE length: %u", ad->data_len);
+
+			return NULL;
+		}
+
+		ltv_len = net_buf_simple_pull_u8(&net_buf);
+		if (net_buf.len < ltv_len) {
+			LOG_DBG("Invalid BASE length: %u", ad->data_len);
+
+			return NULL;
+		}
+		net_buf_simple_pull_mem(&net_buf, ltv_len);
+
+		for (uint8_t j = 0U; j < bis_count; j++) {
+			uint8_t bis_index;
+
+			if (net_buf.len < sizeof(bis_index)) {
+				LOG_DBG("Invalid BASE length: %u", ad->data_len);
+
+				return NULL;
+			}
+
+			bis_index = base_pull_bis_count(&net_buf);
+			if (bis_index == 0 || bis_index > BT_ISO_BIS_INDEX_MAX) {
+				LOG_DBG("Subgroup[%u]: Invalid BIS index: %u", i, bis_index);
+
+				return NULL;
+			}
+
+			/* Pull BIS CC data */
+			if (net_buf.len < sizeof(ltv_len)) {
+				LOG_DBG("Invalid BASE length: %u", ad->data_len);
+
+				return NULL;
+			}
+
+			ltv_len = net_buf_simple_pull_u8(&net_buf);
+			if (net_buf.len < ltv_len) {
+				LOG_DBG("Invalid BASE length: %u", ad->data_len);
+
+				return NULL;
+			}
+			net_buf_simple_pull_mem(&net_buf, ltv_len);
+		}
+	}
+
+	return base;
+}
+
+int bt_bap_base_get_pres_delay(const struct bt_bap_base *base)
+{
+	struct net_buf_simple net_buf;
+	uint32_t pd;
+
+	CHECKIF(base == NULL) {
+		LOG_DBG("base is NULL");
+
+		return -EINVAL;
+	}
+
+	net_buf_simple_init_with_data(&net_buf, (void *)base, sizeof(pd));
+	pd = base_pull_pd(&net_buf);
+
+	return (int)pd; /* PD is 24-bit so it fits in an int */
+}
+
+int bt_bap_base_get_subgroup_count(const struct bt_bap_base *base)
+{
+	struct net_buf_simple net_buf;
+	uint8_t subgroup_count;
+
+	CHECKIF(base == NULL) {
+		LOG_DBG("base is NULL");
+
+		return -EINVAL;
+	}
+
+	net_buf_simple_init_with_data(&net_buf, (void *)base, BASE_MAX_SIZE);
+	base_pull_pd(&net_buf);
+	subgroup_count = net_buf_simple_pull_u8(&net_buf);
+
+	return (int)subgroup_count; /* subgroup_count is 8-bit so it fits in an int */
+}
+
+int bt_bap_base_foreach_subgroup(const struct bt_bap_base *base,
+				 bool (*func)(const struct bt_bap_base_subgroup *data,
+					      void *user_data),
+				 void *user_data)
+{
+	struct bt_bap_base_subgroup *subgroup;
+	struct net_buf_simple net_buf;
+	uint8_t subgroup_count;
+
+	CHECKIF(base == NULL) {
+		LOG_DBG("base is NULL");
+
+		return -EINVAL;
+	}
+
+	CHECKIF(func == NULL) {
+		LOG_DBG("func is NULL");
+
+		return -EINVAL;
+	}
+
+	net_buf_simple_init_with_data(&net_buf, (void *)base, BASE_MAX_SIZE);
+	base_pull_pd(&net_buf);
+	subgroup_count = net_buf_simple_pull_u8(&net_buf);
+
+	for (uint8_t i = 0U; i < subgroup_count; i++) {
+		subgroup = (struct bt_bap_base_subgroup *)net_buf.data;
+		if (!func(subgroup, user_data)) {
+			LOG_DBG("user stopped parsing");
+
+			return -ECANCELED;
```

Comment:
I'm not convinced this should result in error

Reply:
I'm using the same approach as I did for `bt_audio_data_parse`. 

I think it is useful for the caller to know if it was prematurely cancelled, although looking at both `bt_audio_data_parse` and `bt_bap_base_foreach_subgroup`, there we do not have any case where we change behavior based on it. 

If you like, I can remove it and just return `0` (similar to `bt_data_parse`).

Reply:
Got it